akka-http-core/src/main/scala/akka/http/impl/engine/http2/ResponseRendering.scala

```diff
@@ -5,17 +5,61 @@
 package akka.http.impl.engine.http2
 
 import akka.event.LoggingAdapter
+import akka.http.impl.engine.http2.FrameEvent.ParsedHeadersFrame
 import akka.http.impl.util.StringRendering
 import akka.http.scaladsl.model._
 import akka.http.scaladsl.model.headers.Date
 import akka.http.scaladsl.settings.ServerSettings
 
 import scala.collection.immutable
 import scala.collection.immutable.VectorBuilder
-import FrameEvent.ParsedHeadersFrame
 
 private[http2] object ResponseRendering {
 
+  def renderResponse(settings: ServerSettings, log: LoggingAdapter): HttpResponse => Http2SubStream = {
+    def failBecauseOfMissingAttribute: Nothing =
+      // attribute is missing, shutting down because we will most likely otherwise miss a response and leak a substream
+      // TODO: optionally a less drastic measure would be only resetting all the active substreams
+      throw new RuntimeException("Received response for HTTP/2 request without x-http2-stream-id attribute. Failing connection.")
+
+    new CommonRendering[HttpResponse] {
+      override def nextStreamId(response: HttpResponse): Int = response.attribute(Http2.streamId).getOrElse(failBecauseOfMissingAttribute)
+
+      override def initialHeaderPairs(response: HttpResponse): VectorBuilder[(String, String)] = {
+        val headerPairs = new VectorBuilder[(String, String)]()
+        // From https://tools.ietf.org/html/rfc7540#section-8.1.2.4:
+        //   HTTP/2 does not define a way to carry the version or reason phrase
+        //   that is included in an HTTP/1.1 status line.
+        headerPairs += ":status" -> response.status.intValue.toString
+      }
+
+      override lazy val peerIdHeader: Option[(String, String)] = settings.serverHeader.map(h => h.lowercaseName -> h.value)
+    }.createRenderer(log)
+
+  }
+}
+
+trait CommonRendering[R <: HttpMessage] {
+
+  def nextStreamId(r: R): Int
+  def initialHeaderPairs(r: R): VectorBuilder[(String, String)]
+  def peerIdHeader: Option[(String, String)]
+
+  def createRenderer(log: LoggingAdapter): R => Http2SubStream = { (r: R) =>
+
+    val headerPairs = initialHeaderPairs(r)
+
+    CommonRendering.addContentHeaders(headerPairs, r.entity)
+    CommonRendering.renderHeaders(r.headers, headerPairs, peerIdHeader, log, isServer = r.isResponse)
+
+    val headersFrame = ParsedHeadersFrame(nextStreamId(r), endStream = r.entity.isKnownEmpty, headerPairs.result(), None)
+
+    Http2SubStream(r.entity, headersFrame, r.attributes.filter(_._2.isInstanceOf[RequestResponseAssociation]))
+  }
+}
+
+private[http2] object CommonRendering {
+
```

Comment:
I first moved a few methods and fields into the new `CommonRendering` and later created the `CommonRendering` trait.

I think the current names for the object and the trait are not grat but keeping some code on the `object` came in quite handy on tests.